CGMES : Implement a pre-processor that completes the CGMES model with missing containers before starting the conversion

zamarrenolm opened the pull request:
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#2458 


**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature


**What is the current behavior?** *(You can also link to an open issue here)*
CGMES import fails if containers (voltage levels) are referred but not defined in the set of input XML files.


**What is the new behavior (if this is a feature change)?**
A CGMES pre-processor scheme is proposed, following a similar approach to current post-processors.
Here we want to add plugins that intervene in the CGMES model AFTER the given CGMES data has been read but BEFORE the conversion is started.
This provides a point where we can configure a potentially time consuming process. It will be called only if the user activates it.
We inspect the initial CGMES model looking for missing data and try to complete it before the conversion starts.

**Other information**:
The implementation will build a new XML file that will contain the missing definitions.
After reading it to add its contents to the CGMES model in memory, we have to invalidate all previous caches.

miovd commented on cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/DefineMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,91 @@
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ */
+@AutoService(CgmesImportPreProcessor.class)
+public class DefineMissingContainersPreProcessor implements CgmesImportPreProcessor {
```
To be discussed: not sure we want this pre-processor in cgmes-conversion. In a new module, cgmes-validation?

zamarrenolm replied:
I prefer the option of using a separate module instead of a different repo, as it would keep closer all the cgmes-related code.

I would like a name that, in addition to "validation", suggests that we could also be "fixing" the model ... cgmes-fixes ? cgmes-completion ?

zamarrenolm replied:
I've moved this pre-processor to a new module named cmges-completion.

miovd commented on cgmes/cgmes-model/src/main/resources/CIM16.sparql:
```diff
@@ -264,7 +264,8 @@ SELECT *
 WHERE {
-{ GRAPH ?graphEQ {
+# FIXME(Luma) when we fix voltage levels VL definitions may be in a different graph from the original Bay
+# { GRAPH ?graphEQ {
```
Do we want to modify this file or introduce a new file with less strict requirements?

zamarrenolm replied:
In this case we could directly modify the file, as we are just iterating over all connectivity node containers in all graphs.

We generally preserve graph clauses to limit the search of related data and improve performance. Here we were only implicitly enforcing that bays must be in the same graph of its voltage level, and removing the graph clause should not have a big impact. To be measured, anyway.

zamarrenolm replied:
Ok to remove the GRAPH clause, and we will keep a comment that refers that the attributes queried in this clause are found in the EQ instance files.

miovd reviewed (commented):
Some remarks to open a discussion

zamarrenolm commented:
When no folder is specified as parameter, the completion files are written in a temp directory created from the current `user.dir` to avoid the security risk defined in https://rules.sonarsource.com/java/RSPEC-5443.

miovd commented on cgmes/cgmes-completion/pom.xml:
```diff
@@ -0,0 +1,97 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2017, RTE (http://www.rte-france.com)
```
*copyright 2023

zamarrenolm replied:
updated

miovd commented on cgmes/cgmes-completion/src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,26 @@
+    <logger name="com.powsybl.triplestore.TripleStore" level="error"/>
+    <logger name="com.powsybl.cgmes.triplestore.CgmesModelTripleStore" level="error"/>
+    <logger name="com.powsybl.cgmes.conversion.Conversion" level="error"/>
+    <logger name="com.powsybl.cgmes.completion" level="info"/>
```
Why is the log level info here?

zamarrenolm replied:
For the unit tests, to have easy access to the relevant info about the run (location of the output file, list of missing voltage levels in the input) ...

zamarrenolm replied:
log level updated to `error`

miovd commented on cgmes/cgmes-conversion/src/test/java/com/powsybl/cgmes/conversion/CgmesImportPostProcessorTest.java:
```diff
@@ -28,6 +29,7 @@
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
+// FIXME(Luma) rename to CgmesImportProcessorsTest (we will test both pre- and post- processors)
```
You can rename it now :)

zamarrenolm replied:
Yes, I forgot to do it. I have renamed the test class as `CgmesImportPreAndPostProcessorsTest`

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/DefineMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ */
+@AutoService(CgmesImportPreProcessor.class)
+public class DefineMissingContainersPreProcessor implements CgmesImportPreProcessor {
```
Maybe `CreateMissingContainersPreProcessor` is more explicit?

zamarrenolm replied:
done

miovd reviewed (commented):
I'm wondering if this pre-processor's place is in core at all.... The feature of CGMES pre-processors in itself is very nice and a good idea! But this one is, I think, about a use-case that is too specific to be in core. Maybe it can be in another repo? @annetill do you have any opinion?

zamarrenolm commented:
Generic scheme for pre-processors has been moved to #2513. Once it is merged in main, we will leave here only the specific code for completing CGMES model with missing containers using the pre-processor mechanism.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+@AutoService(CgmesImportPreProcessor.class)
+public class CreateMissingContainersPreProcessor implements CgmesImportPreProcessor {
+
+    public static final String NAME = "CreateMissingContainers";
```
For homogenisation purposes, you can simply name it `missingContainers`

zamarrenolm replied:
But for example the IIDM modification post-processor that replaces tie lines by lines starts also with a verb (`replaceTieLinesByLines`). I prefer if we keep the action in the name. What do you think?

miovd replied:
I think there is a difference between modifications (which state the action they do) and processors (which state the context, particularly for CGMES processors) but both are okay in my opinion but it should start with a low case

zamarrenolm replied:
The name now starts with lower case, `createMissingContainers`.
I think of pre/post-processors as specialized tools carrying out specific tasks before/after CGMES import, so for me it is ok if their name starts with a verb.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    private void buildFixesOnFolder(CgmesModel cgmes, String basename, Path fixesFolder) {
```
Can this be static?

zamarrenolm replied:
yes, done.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+
+    private static void buildZipFileWithFixes(Set<String> missingVoltageLevels, Path fixesFile, String basename) {
+        Network network = NetworkFactory.findDefault().createNetwork("empty", "CGMES");
+        // TODO(Luma) should we ensure that context CIM version uses the same version of the input files ?
```
Can you elaborate this `TODO`?

zamarrenolm replied:
We have read CGMES files in a given version.
We should export CGMES files in the same version.
When the network is created applying conversion from CGMES files, the version of input data is stored in the extension `CimCharacteristics`, and when an export context is created from a `Network`, the version to use in the export is the same one that was read, unless explicitly changed.
Here we are creating an export context with an empty `Network` that will not carry information about specific CIM version to use, so the default value will be used (CIM16).

I will set the CIM version in the context explicitly, based on the version of the input files.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+        String substationName = voltageLevelId + "SUB for missing VL " + voltageLevelId;
+        String baseVoltageId = CgmesExportUtil.getUniqueId();
+        // An arbitrary, abnormal value
+        double nominalV = 10000.0;
```
Maybe put it as parametrisable? At minimum, it should be a constant to be clear, not hidden in the code

zamarrenolm replied:
It is not very useful to parametrise it, the missing voltage levels may correspond to different points of the network. I added a constant.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    static class RegionContainers {
+        String subGeographicalRegionId;
+        String geographicalRegionId;
+    }
```
Can be private

zamarrenolm replied:
done

zamarrenolm commented on cgmes/cgmes-completion/src/test/java/com/powsybl/cgmes/completion/CgmesCompletionTest.java:
```diff
@@ -0,0 +1,59 @@
+        Properties importParams = new Properties();
+        importParams.put(CgmesImport.PRE_PROCESSORS, "CreateMissingContainers");
+
+        // The only way to pass the output folder where we want the fixes to be written is to use a config file
+        // Its contents should be:
+        //
+        //   import-export-parameters-default-value:
+        //     iidm.import.cgmes.fixes-for-missing-containers-filename: "/user/working/area/fixes/..."
+        //
+        // For tests, it is not possible to put a reference to a Jimfs folder,
+        // Even if we write the uri of the folder, and read it back as an uri,
+        // the default file system is used to interpret it
+        //
+        // An alternative would be to write in the config a temp folder that we create here in the tests,
+        // But that means writing in a resource file (the actual path of the temp folder), it is dirty.
+        //
+        // Instead, the CGMES completion processor creates itself a temp folder when no parameter is received.
+        // We won't receive directly the file with the fixes, but we do not care.
```
I did not want to use Jimfs directly in the preprocessor: in the rest of PowSyBl core this library is only included in test code. It will add a new dependency in the framework.

The problem is that we can not setup a Jimfs folder inside the test code that is then used in the preprocessor: no way to put a reference in the config file.

zamarrenolm commented:
Removed the "silent" write if the output folder for fixes is not defined in the configuration.
Adjusted the test to use Jimfs in this scenario.
When an output folder for the files containing the fixes is defined in the configuration: if it is an absolute path it is taken as it is. If it is relative, it is assumed to be relative the the config dir. This allows to create a test config on Jimfs and the pre-processor to use it.

miovd commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,245 @@
+import java.util.zip.ZipOutputStream;
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
```
Maybe a (light) javadoc, just to explain what is done in this pre-processor?

zamarrenolm replied:
added a small javadoc

zamarrenolm commented on cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java:
```diff
@@ -0,0 +1,216 @@
+            String prefix = "cgmes-fixes-for-missing-containers";
+            boolean debug = false;
+            try (WorkingDirectory workingDirectory = new WorkingDirectory(current, prefix, debug)) {
+                buildFixesOnFolder(cgmes, basename, workingDirectory.toPath());
```
I was able to solve the issue (and do proper testing) without the need of a explicit additional parameter for `FileSystem`.

When we obtain the folder parameter from config:

- if it is an absolute path the default file system is used.
- If the path is relative, it is assumed relative to the platform config dir.

In the tests, we create a platform config that lives in the Jimfs file system, following the same schema applied in other unit tests.
